src/content/docs/en/guides/middleware.mdx

```diff
@@ -0,0 +1,176 @@
+
+1. Create `src/middleware.js|ts` (Alternatively, you can create `src/middleware/index.js|ts`.)
+
+2. Inside this file, export an [`onRequest()`](#onrequest) function that modifies a [`locals`](#locals) object. This must not be a default export.
+
+```js title="src/middleware.js"
+export function onRequest ({ locals, request }, resolve) {
+	// locals = {}
+};
+```
+
+3. Inside any `.astro` file, access `Astro.locals`
+
+```astro title="src/components/Component.astro"
+---
+const data = Astro.locals;
+---
+
+<p>This {data.property} is from middleware.</p>
+
+```
+
+### Authentication example
+
+Here is an example of an authentication middleware:
+
+```ts
+import type { MiddlewareResponseHandler } from "astro/middleware"
+const auth: MiddlewareResponseHandler = async ({ cookies, locals }, next) => {
+	if (!cookies.has("sid")) {
+		return new Response(null, {
+			status: 405, // Not allowed
+		})
+	}
+	let sessionId = cookies.get("sid")
+	let user = await getUserFromSession(sessionId)
+	if (!user) {
+		return new Response(null, {
+			status: 405, // Not allowed
+		})
+	}
+	locals.user = user
+	return next()
+}
+export { auth as onRequest }
+```
+
+Which you can then access via your Astro components:
+
+```astro
+---
+const { user } = Astro.locals
+---
+<h1>Hello {user.name}</h1>
+```
+
+For more usage examples, please see [the middleware RFC proposal](https://github.com/withastro/roadmap/blob/rfc/stage-3-middleware/proposals/0032-middleware-api.md#middleware-workflow-and-examples)
+
+### Chaining middleware
+
+Multiple middlewares can be joined, in a specified order, using [`sequence()`](#sequence)
+
+```js title="src/middleware.js"
+import {sequence} from "astro/middleware";
+
+function validation() {}
+function auth() {}
+
+export const onRequest = sequence(validation, auth);
+```
+
+```js
+import {sequence} from "astro/middleware";
+
+async function validation(_, next) {
+    console.log("validation request");
+    const response = await next();
+    console.log("validation response");
+    return response;
+}
+async function auth(_, next) {
+  console.log("auth request");
+  const response = await next();
+  console.log("auth response");
+  return response;
+
+}
+async function greeting(_, next) {
+  console.log("greeting request");
+  const response = await next();
+  console.log("greeting response");
+  return response;
+
+}
+
+export const onRequest = sequence(validation, auth, greeting);
+```
+
+will result in the following console order:
+
+```sh
+validation request
+auth request
+greeting request
+greeting response
+auth response
+validation response
+```
+
+## API reference
+
+### onRequest()
+
+A required exported function from `src/middleware.js` that will be called before rendering every page or API route. It accepts two arguments: [context](#context) and [next()](#next)
+
+### context
+
+An object that includes information to be made available to other middleware, API routes and `.astro` routes during the rendering process. 
```

Comment:
This is the same object that is passed to API endpoints. Not sure if that's worth mentioning here or not.